### ISSUE TITLE
django.db.utils.OperationalError: (1101, "BLOB, TEXT, GEOMETRY or JSON column 'attrs' can't have a default value")

### ISSUE THREAD

**Issue by haoziyeung** [adamchainz/django-mysql#417]
Server version: 5.7.20 MySQL Community Server (GPL)
mysqlclient (1.3.12)
Django (1.8.18)

Hello,there i have a problem
i've created a json field like below:
from django_mysql.models import JSONField,ListCharField
class express_model(models.Model):
    '''
        快递公司名单
    '''
    id = models.UUIDField(primary_key=True, default=uuid.uuid4, editable=False)
    exp_code = models.CharField(u'快递公司代码', max_length=10,blank=True, null=True)
    exp_name = models.CharField(u'快递公司名称', max_length=50)
    attrs = JSONField(default=my_default)

then there's ERRORS:

 File "manage.py", line 14, in <module>
    execute_from_command_line(sys.argv)
  File "/usr/lib/python2.7/site-packages/django/core/management/__init__.py", line 354, in execute_from_command_line
    utility.execute()
  File "/usr/lib/python2.7/site-packages/django/core/management/__init__.py", line 346, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/usr/lib/python2.7/site-packages/django/core/management/base.py", line 394, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/usr/lib/python2.7/site-packages/django/core/management/base.py", line 445, in execute
    output = self.handle(*args, **options)
  File "/usr/lib/python2.7/site-packages/django/core/management/commands/migrate.py", line 222, in handle
    executor.migrate(targets, plan, fake=fake, fake_initial=fake_initial)
  File "/usr/lib/python2.7/site-packages/django/db/migrations/executor.py", line 110, in migrate
    self.apply_migration(states[migration], migration, fake=fake, fake_initial=fake_initial)
  File "/usr/lib/python2.7/site-packages/django/db/migrations/executor.py", line 148, in apply_migration
    state = migration.apply(state, schema_editor)
  File "/usr/lib/python2.7/site-packages/django/db/migrations/migration.py", line 115, in apply
    operation.database_forwards(self.app_label, schema_editor, old_state, project_state)
  File "/usr/lib/python2.7/site-packages/django/db/migrations/operations/fields.py", line 62, in database_forwards
    field,
  File "/usr/lib/python2.7/site-packages/django/db/backends/mysql/schema.py", line 43, in add_field
    super(DatabaseSchemaEditor, self).add_field(model, field)
  File "/usr/lib/python2.7/site-packages/django/db/backends/base/schema.py", line 398, in add_field
    self.execute(sql, params)
  File "/usr/lib/python2.7/site-packages/django/db/backends/base/schema.py", line 111, in execute
    cursor.execute(sql, params)
  File "/usr/lib/python2.7/site-packages/django/db/backends/utils.py", line 79, in execute
    return super(CursorDebugWrapper, self).execute(sql, params)
  File "/usr/lib/python2.7/site-packages/django/db/backends/utils.py", line 64, in execute
    return self.cursor.execute(sql, params)
  File "/usr/lib/python2.7/site-packages/django/db/utils.py", line 98, in __exit__
    six.reraise(dj_exc_type, dj_exc_value, traceback)
  File "/usr/lib/python2.7/site-packages/django/db/backends/utils.py", line 64, in execute
    return self.cursor.execute(sql, params)
  File "/usr/lib/python2.7/site-packages/django/db/backends/mysql/base.py", line 124, in execute
    return self.cursor.execute(query, args)
  File "/usr/lib64/python2.7/site-packages/MySQLdb/cursors.py", line 250, in execute
    self.errorhandler(self, exc, value)
  File "/usr/lib64/python2.7/site-packages/MySQLdb/connections.py", line 50, in defaulterrorhandler
    raise errorvalue
django.db.utils.OperationalError: (1101, "BLOB, TEXT, GEOMETRY or JSON column 'attrs' can't have a default value")

Can somebody figure out what happend? is because STRICT MODE?
and here's my settings.py
DATABASES = {
    'default': {
        'ENGINE': 'django.db.backends.mysql',
        'NAME': 'euler_lims',
        'USER':'',
        'PASSWORD':'',
        'HOST':'localhost',
        'PORT':'3306',
        'OPTIONS' : {
            'charset': 'utf8mb4',
            'init_command': "SET sql_mode='STRICT_TRANS_TABLES', innodb_strict_mode=1",
            }
    },
        'TEST': {
            'CHARSET': 'utf8mb4',
            'COLLATION': 'utf8mb4_unicode_ci',
        }
}

**Comment by adamchainz** [comment]
Django doesn't normally set database defaults. Did you edit your migration to include one? Can you capture the `CREATE TABLE` / `ALTER TABLE` statements that are being executed and check them?

**Comment by haoziyeung** [comment]
# -*- coding: utf-8 -*-
from __future__ import unicode_literals

from django.db import migrations, models
import django_mysql.models


class Migration(migrations.Migration):

    dependencies = [
        ('proj', '0002_remove_sample_model_attrs'),
    ]

    operations = [
        migrations.AddField(
            model_name='sample_model',
            name='attrs',
            field=django_mysql.models.JSONField(default=dict),
        ),
    ]

**Comment by haoziyeung** [comment]
[root@985f25c0ab58 eulerlims]# ./manage.py sqlmigrate proj 0002
BEGIN;
ALTER TABLE `proj_reportvariant` ADD COLUMN `attrs` json DEFAULT '{}' NOT NULL;
ALTER TABLE `proj_reportvariant` ALTER COLUMN `attrs` DROP DEFAULT;

COMMIT;

**Comment by haoziyeung** [comment]
sometimes it works,sometimes NOT,i don't know why,

[root@985f25c0ab58 eulerlims]# ./manage.py sqlmigrate proj 0002
BEGIN;
ALTER TABLE `proj_reportvariant` ADD COLUMN `attrs` json DEFAULT '{}' NOT NULL;
ALTER TABLE `proj_reportvariant` ALTER COLUMN `attrs` DROP DEFAULT;

COMMIT;
[root@985f25c0ab58 eulerlims]# ./manage.py sqlmigrate groups_manager 0001|grep properties
CREATE TABLE `groups_manager_group` (`id` integer AUTO_INCREMENT NOT NULL PRIMARY KEY, `name` varchar(255) NOT NULL, `codename` varchar(255) NOT NULL, `description` longtext NOT NULL, `comment` longtext NOT NULL, `full_name` varchar(255) NOT NULL, `properties` json NOT NULL, `django_auth_sync` bool NOT NULL, `lft` integer UNSIGNED NOT NULL, `rght` integer UNSIGNED NOT NULL, `tree_id` integer UNSIGNED NOT NULL, `level` integer UNSIGNED NOT NULL, `django_group_id` integer NULL);

you can see,sometimes SQL have json DEFAULT,,i know json field can't have default value (according mysql manual),sometimes SQL don't have json DEFAULT,then it works

**Comment by haoziyeung** [comment]
i think it's a BUG,you should delete _

**_DEFAULT '{}'_**

_ 
and delete DROP DEFAULT

**Comment by haoziyeung** [comment]
hello,@adamchainz,is this fixed? thanks

**Comment by adamchainz** [comment]
What django version are you using? I just remembered I fixed this in https://code.djangoproject.com/ticket/26966 / https://github.com/django/django/pull/6997 for Django 1.10.

**Comment by BrunoDesthuilliers** [comment]
Same issue here using Django 1.10.6 :-/

**Comment by adamchainz** [comment]
My bad, it's only on 1.11+ as you can see on the tags accessible here: https://github.com/django/django/commit/4b1c9708d629c08ebbd73293c89a159ccb436d4d

Whilst you *should* upgrade to 1.11 anyway, you can always subclass the DB backend and copy in the new version of the function (or cheat and monkey-patch) to fix this on 1.10.

**Comment by BrunoDesthuilliers** [comment]
I managed to get the migration working on Django 1.10 by monkeypatching django.db.backends.mysql.DatabaseSchemaEditor.skip_default in the migration file.

**Comment by adamchainz** [comment]
Share the monkeypatch?

**Comment by BrunoDesthuilliers** [comment]
Well I mostly copy-pasted the diff from 4795936d772304d499e171cb7917891b493a9d80 :

```
# -*- coding: utf-8 -*-
# Generated by Django 1.10.6 on 2018-02-02 16:34
from __future__ import unicode_literals

from django.db import migrations
import django_mysql.models

# cf https://github.com/adamchainz/django-mysql/issues/417 
# this is supposed to be fixed in django 1.11 but in the meantime 
# we need this patch to get the migration working
from django.db.backends.mysql import schema
def skip_default(self, field):
    db_type = field.db_type(self.connection)
    return (
        db_type is not None and
        db_type.lower() in {
            'tinyblob', 'blob', 'mediumblob', 'longblob',
            'tinytext', 'text', 'mediumtext', 'longtext',
            'json',
            }
        )
schema.DatabaseSchemaEditor.skip_default=skip_default

class Migration(migrations.Migration):
    dependencies = [
        ('blookcore', '0028_auto_20180202_1644'),
    ]

    operations = [
        migrations.AddField(
            model_name='blook',
            name='_options',
            field=django_mysql.models.JSONField(blank=True, db_column=b'options', default=dict, editable=False, verbose_name=b'options'),
        ),
    ]
```

**Comment by adamchainz** [comment]
Cool, just useful for anyone googling this in the future. Thanks, I'm gonna close this as we have a documented workaround now.

**Comment by adamchainz** [comment]
It's in this comment: https://github.com/adamchainz/django-mysql/issues/417#issuecomment-363082924

But also I think I got the fix into Django now.

**Comment by debdutgoswami** [comment]
@adamchainz I'm using django 3.2 and MySQL 8.0.13+ but I'm still facing the issue while running the migration

**Comment by DanielSwain** [comment]
Just reporting that this problem is still happening.  I'm working on an old project that I've upgraded to mysql 8.0.35 and Django 4.2.7.  A `CharField` had been defined with `blank=True, null=True`.  I changed it to `blank=True, default=''` and created the migration.  The migration ran fine on SQLite in development, but it failed in MySQL in production.  I tried removing `default=True` from the `AlterField` in the migration as  [described here by Matt Westcott](https://code.djangoproject.com/ticket/32503#:~:text=I%20worked%20around%20this%20by%20splitting%20the%20AlterField%20operation%20into%20two%20steps), but I still got the error message about a default not being allowed.  Since the migration was only setting `default=True`, I ended up and just faked it.  This is really odd though, because when I do `describe table_name` for the table in question, it shows the following descriptive information (note the Default column):

![image](https://github.com/adamchainz/django-mysql/assets/7819909/7ad16239-48f4-4d21-a69b-16d4ca091896)

I even tried `alter table my_table modify name TEXT DEFAULT '';` from the mysql command line, but it still gave the same error:
ERROR 1101 (42000): BLOB, TEXT, GEOMETRY or JSON column 'name' can't have a default value

Just leaving this here for anyone else who comes across this issue.  We use Postgresql for nearly everything these days, so this problem makes me appreciate PG that much more.
